notebooks/how-to/03-adaptivity.ipynb

```diff
@@ -0,0 +1,876 @@
+   "outputs": [],
+   "source": [
+    "fields_2 = run_simulation(\n",
+    "    solver, final_time, dt, bed=z_b, **fields\n",
+    ")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "To get an idea of where we're making the largest errors, we can look at the discrepancy between the degree-1 and degree-2 simulations."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "expr = abs(fields_2['thickness'] - fields_1['thickness'])\n",
+    "δh = interpolate(expr, Q2)\n",
+    "\n",
+    "fig, axes = subplots()\n",
+    "colors = firedrake.tripcolor(δh, axes=axes)\n",
+    "colorbar(fig, colors);"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Perhaps not surprisingly, the biggest misfits occur near the grounding line.\n",
+    "Now that we know where we need more triangles, how do we go about refining the mesh?"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Refinement\n",
+    "\n",
+    "Before actually refining the mesh, we'll want to create a new field that smooths over the thickness error in space.\n",
+    "This will help give a more continuous gradation between coarse and fine triangles rather than a sudden jump."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from firedrake import dS\n",
+    "\n",
+    "DG0 = firedrake.FunctionSpace(coarse_mesh, 'DG', 0)\n",
+    "ϵ = firedrake.Function(DG0)\n",
+    "J = 0.5 * (\n",
+    "    (ϵ - δh)**2 * dx +\n",
+    "    (Ly / 2) * (ϵ('+') - ϵ('-'))**2 * dS\n",
+    ")\n",
+    "F = firedrake.derivative(J, ϵ)\n",
+    "firedrake.solve(F == 0, ϵ)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "fig, axes = subplots()\n",
+    "colors = firedrake.tripcolor(ϵ, axes=axes)\n",
+    "colorbar(fig, colors);"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The `element_volumes` member of the Triangle mesh data structure contains an array that we'll fill in order to specify the desired triangle areas in the refined mesh.\n",
+    "This array isn't initialized by default.\n",
+    "The setup routine below allocates space for it."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "triangle_mesh.element_volumes.setup()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Now we have to make some decisions about how much to actually refine the mesh.\n",
+    "Here we'll specify arbitrarily that the triangles with the largest errors will have their areas shrunk by a factor of 8.\n",
+    "We then have to decide how to much to shrink the areas of triangles with less than the largest error.\n",
```

Comment:
delete "to":  We then have to decide how much to shrink the areas of triangles with less than the largest error.